include/boost/geometry/extensions/io/plotly/plotter.hpp

```diff
@@ -0,0 +1,242 @@
+            y << (first ? "\"" : ", \"" )
+              << geometry::get<1>(*it)
+              << "\"";
+            if( Dimension > 2 ) 
+            {
+                z << (first ? "\"" : ", \"" )
+                  << geometry::get<2>(*it)
+                  << "\"";
+            }
+        }
+
+        x << "]";
+        y << "]";
+        z << ( Dimension > 2 ? "]" : "" );
+        
+        os << x.rdbuf()
+           << y.rdbuf() 
+           << z.rdbuf()
+           << ", \"line\": {\"dash\": \"solid\", \"color\": \"" << color << "\", \"width\": 2}";
+    }
+};
+
+
+template <typename MultiGeometry, typename Policy>
+struct json_multi
+{
+    template <typename Char, typename Traits>
+    static inline void apply(std::basic_ostream<Char, Traits>& os,
+        MultiGeometry const& multi, std::string const& color)
+    {
+        for (typename boost::range_iterator<MultiGeometry const>::type
+                    it = boost::begin(multi);
+            it != boost::end(multi);
+            ++it)
+        {
+            Policy::apply(os, *it, color);
+        }
+
+    }
+
+};
+
+}} // namespace detail::json
+#endif // DOXYGEN_NO_DETAIL
+
+#ifndef DOXYGEN_NO_DISPATCH
+namespace dispatch
+{
+
+template 
+<
+    typename Geometry, 
+    typename Tag = typename tag<Geometry>::type
+>
+struct json_plot
+{
+    BOOST_MPL_ASSERT_MSG
+        (
+            false, NOT_OR_NOT_YET_IMPLEMENTED_FOR_THIS_GEOMETRY_TYPE
+            , (Geometry)
+        );
+};
+
+template <typename Point>
+struct json_plot<Point, point_tag> : detail::json::json_point<Point, dimension<Point>::value> {};
+
+template <typename Segment>
+struct json_plot<Segment, segment_tag> : detail::json::json_segment<Segment, dimension<Segment>::value> {};
+
+template <typename Linestring>
+struct json_plot<Linestring, linestring_tag>
+    : detail::json::json_range<Linestring,  dimension<Linestring>::value> {};
+
+template <typename MultiPoint>
+struct json_plot<MultiPoint, multi_point_tag>
+    : detail::json::json_multi
+        <
+            MultiPoint,
+            detail::json::json_point
+                <
+                    typename boost::range_value<MultiPoint>::type,
+                    dimension<MultiPoint>::value
+                >
+        >
+{};
+
+template <typename MultiLinestring>
+struct json_plot<MultiLinestring, multi_linestring_tag>
+    : detail::json::json_multi
+        <
+            MultiLinestring,
+            detail::json::json_range
+                <
+                    typename boost::range_value<MultiLinestring>::type,
+                    dimension<MultiLinestring>::value
+                >
+
+        >
+{};
+
+} // namespace dispatch
+#endif // DOXYGEN_NO_DISPATCH
+
+template <typename Geometry>
+inline void json_plot(std::ostream& stream, Geometry const& geometry, std::string const& color)
+{
+    dispatch::json_plot<Geometry>::apply(stream, geometry, color);
+}
+
+
+class json_plotter : boost::noncopyable
+{
+    std::ostream& m_stream;
```

Comment:
Ok, so the stream has to live outside this class. It is like in `svg_mapper` so I understand that you replicated the existing design. This is error-prone since when the object outside is destroyed then this is a dangling reference.

So I'm wondering about a different design, i.e. owning the stream instead of the reference. E.g.:
```
template <typename Stream>
class plotly_plotter
{
public:
    template <typename ...Ts>
    plotly_stream(Ts &&... params)
        : m_stream(std::forward<Ts>(params)...)
    {}
private:
    Stream m_stream;
};
```

What do you think? Do you have different ideas?

Reply:
Yeah that seems better. I was thinking why the above design was implemented in `svg_mapper`. The only reason I could find is if some external changes are required by the user that isn't supported by `svg_mapper`. This isn't required by plotly plotter at all then.  Can you tell me that actual reason for this design?

Reply:
> Can you tell me that actual reason for this design?

I don't really know I have only a few ideas though. @barendgehrels do you remember why exactly does the `svg_mapper` store the reference to the stream instead of owning it?

Back then perfect forwarding was impossible. The code in my previous comment requires C++11 but Boost.Geometry right now requires C++03 (this will change soon). And various streams can have various interfaces, e.g. no `open()` member function so there must be a way of passing parameters into its constructor.

In some cases the stream objects are already created, e.g. `std::cout`, `std::cerr`, etc. AFAIU it is possible to account for that by creating a `std::ostream` from another stream's `streambuf`:
```
plotly_plotter<std::ostream> plotter(std::cout.rdbuf());
```
But this would be more confusing than simply calling:
```
plotly_plotter plotter(std::cout);
```

With non-owning relationship the same stream can be shared by several mappers/plotters. I don't know if that makes sense but it is possible nevertheless.

The user can keep and e.g. close and reopen `ofstream` buffer instead of creating the `svg_mapper`/`plotly_plotter` object each time.

> The only reason I could find is if some external changes

Right, though the owned stream could be exposed with some getter function if needed.

Btw, I'm not saying that you should start implementing what I have proposed above. I'm only brainstorming. E.g. now after listing the potential reasons above I'm not sure whether or not my proposal was good.

On the other hand maybe passing a reference as a template parameter could work with the design I proposed. So the user would be able to do both things, store a stream object or a reference. Unless there are some issues I'm not taking into account.
```
plotly_plotter<std::ostream &> plotter(std::cout);
```

And last but not least I'm not sure if the design should be different than `svg_mapper` because the users may expect that the tools work the same and either expect that a reference is stored in `plotly_plotter` or that an object is stored in `svg_mapper`. We should probably be consistent one way or another.

@mloskot do you have some thoughts about it?

Reply:
For now I have put it inside the class itself as I think it's better to make it safe than just copy the existing design as:

1) It is an extension so the design need not be matched with integral parts of Boost Geometry.
2) I see no possible requirement for keeping it outside other than the one mentioned above as we are going to shift to C++11 soon.

Reply:
Although I can't remember exactly, I think @awulkiew has figured it out right.
It was to allow user to plug in a stream object of their choice, and with simple interface.